core/shared/src/main/scala/fs2/util/Concurrent.scala

```diff
@@ -0,0 +1,299 @@
+
+    /** Like modify, but allows to extra `b` in single step. **/
+    def modify2[B](f: A => (A,B)): F[(Change[A], B)] =
+      tryModify2(f).flatMap {
+        case None => modify2(f)
+        case Some(changeAndB) => F.pure(changeAndB)
+      }
+
+    /**
+     * *Asynchronously* sets a reference. After the returned `F[Unit]` is bound,
+     * the task is running in the background. Multiple tasks may be added to a
+     * `Ref[A]`.
+     *
+     * Satisfies: `r.set(t) flatMap { _ => r.get } == t`.
+     */
+    def set(a: F[A]): F[Unit]
+
+    /**
+     * *Asynchronously* sets a reference to a pure value.
+     *
+     * Satisfies: `r.setPure(a) flatMap { _ => r.get(a) } == pure(a)`.
+     */
+    def setPure(a: A): F[Unit] = set(F.pure(a))
+
+    /**
+     * Runs `f1` and `f2` simultaneously, but only the winner gets to
+     * `set` to this `ref`. The loser continues running but its reference
+     * to this ref is severed, allowing this ref to be garbage collected
+     * if it is no longer referenced by anyone other than the loser.
+     */
+    def race(f1: F[A], f2: F[A]): F[Unit]
+  }
+
+  /**
+   * The result of a `Ref` modification. `previous` is the value before modification
+   * (the value passed to modify function, `f` in the call to `modify(f)`. `now`
+   * is the new value computed by `f`.
+   */
+  final case class Change[+A](previous: A, now: A) {
+    def modified[AA >: A](implicit eq: Eq[AA]): Boolean = eq.neqv(previous, now)
+    def map[B](f: A => B): Change[B] = Change(f(previous), f(now))
+  }
+
+  /** Constructs a `Concurrent` instance from a supplied `Effect` instance and execution context. */
+  def mk[F[_]](F: Effect[F])(implicit ec: ExecutionContext): Concurrent[F] = {
+    final class MsgId
+    sealed abstract class Msg[A]
+    object Msg {
+      final case class Read[A](cb: Attempt[(A, Long)] => Unit, id: MsgId) extends Msg[A]
+      final case class Nevermind[A](id: MsgId, cb: Attempt[Boolean] => Unit) extends Msg[A]
+      final case class Set[A](r: Attempt[A]) extends Msg[A]
+      final case class TrySet[A](id: Long, r: Attempt[A], cb: Attempt[Boolean] => Unit) extends Msg[A]
+    }
+
+    new Concurrent[F] { self =>
+      def pure[A](a: A): F[A] = F.pure(a)
+      def handleErrorWith[A](fa: F[A])(f: Throwable => F[A]): F[A] = F.handleErrorWith(fa)(f)
+      def raiseError[A](e: Throwable): F[A] = F.raiseError(e)
+      def async[A](k: (Either[Throwable, A] => Unit) => Unit): F[A] = F.async(k)
+      def runAsync[A](fa: F[A])(cb: Either[Throwable, A] => IO[Unit]): IO[Unit] = F.runAsync(fa)(cb)
+      def flatMap[A, B](fa: F[A])(f: A => F[B]): F[B] = F.flatMap(fa)(f)
+      def tailRecM[A, B](a: A)(f: A => F[Either[A, B]]): F[B] = F.tailRecM(a)(f)
+      def liftIO[A](ioa: IO[A]): F[A] = F.liftIO(ioa)
+      def suspend[A](thunk: => F[A]): F[A] = F.suspend(thunk)
+      def unsafeRunAsync[A](fa: F[A])(f: Either[Throwable, A] => IO[Unit]): Unit =
+        F.runAsync(F.shift(fa)(ec))(f).unsafeRunSync
```

Comment:
Yeah, I'm not in love with the name -- it shifts the computation to the captured EC and then runs it. It's more than just convenience b/c we don't pass ECs throughout the Concurrency `carrying` functions.

An alternative design is making `Concurrency` a utility instead of a type class. Then we'd have to thread `Async[F]` and `ExecutionContext` through things.